docs(feat): add guide how to implement a new per-route feature (#452)
cloudfoundry/routing-release: docs/03-how-to-add-new-route-option.md

```
applications:
- name: test
  routes:
  - route: example.com
    options:
      loadbalancing: round-robin
  - route: example2.com
    options:
      loadbalancing: least-connection
```

**NOTE**: In the implementation, the `options` property of a route represents per-route features.

## Overview
The picture provides a simplified overview of the participating components. The components marked in green need to be enhanced to support additional per-route features.

![](images/components.png)

Implementing per-route features in Cloud Foundry is a laborious process because it requires changes to multiple components in several working groups, namely Application Platform Runtime and Application Runtime Interface.

## Required Changes

* Create an RFC (or an amendment to the existing one) to communicate a new per-route feature
* Create an implementation issue similar to [tracking issue #909](https://github.com/cloudfoundry/community/issues/909)
* Find your use case and implement

### Use Cases
#### New Feature for Application Routes 
To introduce a new application per-route option, follow the instructions:

> * Extend **Cloud Controller** to accept and process a new per-route options in the manifest within [app/messages](https://github.com/sap-contributions/cloud_controller_ng/tree/main/app/messages), especially `route-*-message.rb ` and `manifest_routes_*_message.rb`.

Reviewer: ```suggestion
* Extend **Cloud Controller** to accept and process a new per-route options in the manifest and the API. The [Pull-Request](https://github.com/cloudfoundry/cloud_controller_ng/pull/4080/files) which introduced per-route options with the `loadbalancing` option can serve as a starting point to determine required changes. Update the documentation in the `docs` folder accordingly.
```

* There is no need to adapt the coding in **BBS**. The `route` object is stored as a generic JSON object in BBS. Therefore, BBS simply accepts the `options` it receives from Cloud Controller and saves them as a string in its database. For more details, refer to the discussion on [BBS issue #939](https://github.com/cloudfoundry/diego-release/issues/939).
* There is no need to change anything in **route-emitter** and **routing-info** components. The initial implementation of per-route features extended the `CFRoute` with `options` as a raw JSON message. The only additional step you might consider is implementing tests for these components to ensure that the new option is included in the raw JSON.

> * There is no need to implement anything in **NATS**, as NATS only forwards the route registration messages.

Reviewer: Maybe also mention that there is no need for a change in the CLI, as it takes an arbitrary list of key-value pairs (which it forwards to cloud controller).

Author: Applied!

* Implement your logic in **Gorouter** if it does not already exist. Extend the options included in the registration message within the [mbus/subscriber](https://github.com/cloudfoundry/gorouter/blob/b0d88bb6204cf28e476b4ee680a6f5a154885608/mbus/subscriber.go#L1). The structure `RegistryMessageOpts` represents the property `options` of the NATS registration message. This property can contain additional, custom configuration for a route.

> * Please do not forget adapt examples and extend the documentation related to your changes.

Reviewer: ```suggestion
* Please do not forget to adapt examples and to extend the documentation related to your changes.
```

---
title: How To Add a New Route Option
expires_at: never
tags: [routing-release]
---

## What are Per-Route Features?
Before this feature was implemented, the Cloud Foundry routing stack did not support configuring features for specific routes. Most features could only be defined at the platform level. The generic per-route features allow for defining specific options on a per-route basis. 

> The introduction of the generic per-route features was detailed in [RFC-0027](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0027-generic-per-route-features.md) and referenced in the [community implementation issue](https://github.com/cloudfoundry/community/issues/909). The implementation became available since routing-release/v0.329.

Reviewer 2: This should probably also include the CC API version.

Author: Good point!

The first per-route feature implemented is the load balancing algorithm, which defines how the load is distributed between Gorouters and backends. The algorithm can be configured on the route level via the application manifest:
```
applications:
- name: test
  routes:
  - route: example.com
    options:
      loadbalancing: round-robin
  - route: example2.com
    options:

>       loadbalancing: least-connection

Reviewer 2: ```suggestion
      loadbalancing: least-connections
```

```

**NOTE**: In the implementation, the `options` property of a route represents per-route features.

## Overview
The picture provides a simplified overview of the participating components. The components marked in green need to be enhanced to support additional per-route features.

> ![](images/components.png)

Reviewer 2: The image doesn't load for me, I think the path needs to be relative.

```suggestion
![](./images/components.png)
```

Author: I can see the picture in the online version of document. I think the relative path is correct too.

Reviewer 2: Seems like it's only broken in the Review tab.

Implementing per-route features in Cloud Foundry is a laborious process because it requires changes to multiple components in several working groups, namely Application Platform Runtime and Application Runtime Interface.

## Required Changes

* Create an RFC (or an amendment to the existing one) to communicate a new per-route feature
* Create an implementation issue similar to [tracking issue #909](https://github.com/cloudfoundry/community/issues/909)

> * Find your use case and implement

Reviewer 2: I'd say finding (or rather identifying) the use-case is the first step.
